File: linked_list/scedule.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Scedule{
    int time; // 24시간 표시법 사용
    char description[50];
    Scedule *next = NULL;
};

void makeScedule(Scedule** myScedule, int time, char* description);
void fixScedule(Scedule* myScedule, int time, char* change);
void deleteScedule(Scedule** myScedule, int time);
void printScedule(Scedule* myScedule);
void removeAllScedule(Scedule** myScedule);

int main(){
    int choice, isContinued = 1;
    int time;
    char description[50];
    Scedule* saveScedule = NULL;
    while(isContinued){
        printf("==스케쥴 관리 프로그램==\n");
        printf("1. 스케쥴 추가\n");
        printf("2. 스케쥴 수정\n");
        printf("3. 스케쥴 삭제\n");
        printf("4. 스케쥴 출력\n");
        printf("(이외 값 입력시 종료)\n");
        printf(" >> ");
        scanf("%d", &choice);
        switch (choice){
            case 1:
                printf("스케쥴을 등록할 시간대를 입력하세요\n >> ");
                scanf("%d", &time);
                getchar();
                printf("스케쥴의 상세 내용을 입력해주세요.\n >> ");
                fgets(description, sizeof(description), stdin);
                description[strlen(description)-1]='\0';
                makeScedule(&saveScedule, time, description);
                break;
            case 2:
                printf("스케쥴을 수정할 시간대를 입력하세요\n >> ");
                scanf("%d", &time);
                getchar();
                printf("수정할 스케쥴의 상세 내용을 입력해주세요.\n >> ");
                fgets(description, sizeof(description), stdin);
                description[strlen(description)-1]='\0';
                fixScedule(saveScedule, time, description);
                break;
            case 3:
                printf("스케쥴을 삭제할 시간대를 입력하세요\n >> ");
                scanf("%d", &time);
                deleteScedule(&saveScedule, time);
                break;
            case 4:
                printf("===현재 저장되어있는 스케쥴===\n");
                printScedule(saveScedule);
                break;
            default:
                isContinued = 0;
        }
    }
    removeAllScedule(&saveScedule);
}

void makeScedule(Scedule** myScedule, int time, char* description){
    Scedule* newScedule = (Scedule*)malloc(sizeof(Scedule));
    newScedule->time = time;
    strcpy(newScedule->description, description);
    newScedule->next = NULL;
    if(*myScedule == NULL){
        *myScedule = newScedule;
    } else if ((*myScedule)->time > time){
        newScedule->next = (*myScedule)->next;
        *myScedule = newScedule;
    } else if ((*myScedule)->time == time){
        printf("[ERROR] 같은 시간대에 스케쥴을 두개 이상 만들 수 없습니다.\n");
        return free(newScedule);
    } else {
        while((*myScedule)->next){
            if ((*myScedule)->next->time == time){
                printf("[ERROR] 같은 시간대에 스케쥴을 두개 이상 만들 수 없습니다.\n");
                return free(newScedule);
            }
            if((*myScedule)->next->time > time){
                newScedule->next = (*myScedule)->next;
                (*myScedule)->next = newScedule;
                break;
            }
            myScedule = &((*myScedule)->next);
        }
        if((*myScedule)->next == NULL){
            (*myScedule)->next = newScedule;
        }
    }
}

void fixScedule(Scedule* myScedule, int time, char* change){
    if(myScedule==NULL || myScedule->time > time){
        printf("[ERROR] 수정하고 싶은 시간대에 스케쥴이 없습니다.\n");
        return;
    }
    if(myScedule->time == time){
        strcpy(myScedule->description, change);
        return;
    }
    return fixScedule(myScedule->next, time, change);
}

void deleteScedule(Scedule** myScedule, int time){
    Scedule* delScedule;
    if(*myScedule==NULL){
        printf("[ERROR] 현재 스케쥴이 비어있습니다.\n");
        return;
    }
    if((*myScedule)->time==time){
        delScedule = *myScedule;
        *myScedule = (*myScedule)->next;
        return free(delScedule);
    }
    while((*myScedule)->next){
        if((*myScedule)->next->time>time){
            printf("[ERROR] 삭제할 시간대에 스케쥴이 없습니다.\n");
            return;
        }
        if((*myScedule)->next->time==time){
            delScedule = (*myScedule)->next;
            (*myScedule)->next = delScedule->next;
            return free(delScedule);
        }
        myScedule = &((*myScedule)->next);
    }
    printf("[ERROR] 삭제할 시간대에 스케쥴이 없습니다.\n");
}

void printScedule(Scedule* myScedule){
    if(myScedule == NULL) return;
    printf("[%3d시 : %-50s]\n", myScedule->time, myScedule->description);
    return printScedule(myScedule->next);
}

void removeAllScedule(Scedule** myScedule){
    if(*myScedule){
        removeAllScedule(&((*myScedule)->next));
        free(*myScedule);
        *myScedule = NULL;
    }
}

File: linked_list/scedule.py
class Scedule:
    def __init__(self, time, desc):
        self.time = time
        self.desc = desc
        self.next = None

def makeScedule(s, time, desc):
    newScedule = Scedule(time, desc);
    if (s[0]==None):
        s[0]=newScedule
        return
    if (s[0].time > time):
        newScedule.next = s[0];
        s[0] = newScedule;
    elif (s[0].time == time):
        print("[ERROR] 같은 시간대에 스케쥴을 두개 이상 만들 수 없습니다.")
        return
    else:
        tmp = s[0]
        while tmp.next:
            if (tmp.next.time == time):
                print("[ERROR] 같은 시간대에 스케쥴을 두개 이상 만들 수 없습니다.")
                return
            if (tmp.next.time > time):
                newScedule.next = tmp.next;
                tmp.next = newScedule;
                break
            tmp = tmp.next
        if(tmp.next == None):
            tmp.next = newScedule

def fixScedule(s, time, change):
    if(s[0]==None or s[0].time > time):
        print("[ERROR] 수정하고 싶은 시간대에 스케쥴이 없습니다.\n")
        return
    if(s[0].time == time):
        s[0].desc = change
        return
    return fixScedule([s[0].next], time, change)

def deleteScedule(s, time):
    if(s[0]==None):
        print("[ERROR] 현재 스케쥴이 비어있습니다.")
        return
    if(s[0].time == time):
        s[0] = s[0].next
        return
    tmp = s[0]
    while(tmp.next):
        if(tmp.next.time > time):
            print("[ERROR] 삭제할 시간대에 스케쥴이 없습니다.")
            return
        if(tmp.next.time==time):
            tmp.next = tmp.next.next
            return
        tmp = tmp.next

def printScedule(s):
    if(s[0] == None):
        return
    print("[{}시 : {}]".format(s[0].time, s[0].desc))
    return printScedule([s[0].next])

saveScedule = [None]
while(True):
    print("\n==스케쥴 관리 프로그램==")
    print("1. 스케쥴 추가")
    print("2. 스케쥴 수정")
    print("3. 스케쥴 삭제")
    print("4. 스케쥴 출력")
    print("(이외 값 입력시 종료)")
    choice = input(" >> ")
    if(choice=="1"):
        time = int(input("스케쥴을 등록할 시간대를 입력하세요.\n >> "))
        desc = input("스케쥴의 상세 내용을 입력해주세요.\n >> ")
        makeScedule(saveScedule, time, desc)
    elif(choice=="2"):
        time = int(input("스케쥴을 수정할 시간대를 입력하세요.\n >> "))
        desc = input("수정할 스케쥴의 상세 내용을 입력해주세요.\n >> ")
        fixScedule(saveScedule, time, desc)
    elif (choice == "3"):
        time = int(input("스케쥴을 삭할 시간대를 입력하세요.\n >> "))
        deleteScedule(saveScedule, time)
    elif (choice == "4"):
        print("===현재 저장되어있는 스케쥴===")
        printScedule(saveScedule)
    else:
        break